### ISSUE TITLE
FOQUS Optimization:  Failure Handling 

### ISSUE THREAD

**Issue by boverhof** [CCSI-Toolset/FOQUS#696]
# Description
The current Optimization Under Uncertainty (OUU) in FOQUS throws out an entire objective function evaluation if a single scenario fails.  In OUU the objective function is basically a weighted average of scenario objective function value obtained from a scenario optimization performed by Aspen.  If one of the scenario optimizations in Aspen fail, the current behavior is just to consider the whole FOQUS objective function evaluation (sum of scenarios) to be a failure.

The problem is that getting a reasonable objective function value with partial data is difficult. But it is also wasteful to simply discard a lot of successful runs.  As the number of scenarios increase, it becomes increasingly likely that very few objective evaluation will be completely successful which would make OUU impossible.

# Solution

The most obvious solution to this problem is to just omit failed scenarios from the objective evaluation.  This does not work, since the optimization tends to result in solutions where bad scenarios always fail and are removed from the problem.

The second fairly easy solution is to replace the objective value for failed scenarios either by are constants bad value to penalize failed scenarios or some other value generated by historical data for the failed scenario.  This is what we'll try.  Based on how this work we'll decide where to go next.

**Comment by eslickj** [comment]
@fsoepyan, do you think you can take a shot at writing some objective functions based on the above description.  If FOQUS itself needs to be modified to accommodate, let me know.

**Comment by fsoepyan** [comment]
@eslickj: For the objective function, I was thinking something like this:

If running scenario i = success: COE(i) = the output from ACM
If running scenario i = failure: COE(i) = 1000
Objective = sum COE(i) for i = 1 to the total number of scenarios

Is this a reasonable objective function?

**Comment by eslickj** [comment]
Yeah, that's totally reasonable.  I'm not sure what to expect, so let's just see what happens.  I'm pretty sure I tried this and things when bad, but I can't say for sure why, so I think we should try it again and go from there to figure out what to do next.

I guess one possibility is is that penalizing failed scenarios like this may be pretty similar to just saying the whole thing failed.  If total failure is 1000.  Then one scenario failing is 1000 + the other scenarios.

I guess we'll try a bunch of ideas and hopefully one will work out.

**Comment by fsoepyan** [comment]
@eslickj: I talked with @omellben, and his idea is this:

1. For each iteration, we run all the scenarios.
2. We compute the value of the COE at each scenario.
3. For the successful scenarios, we determine the maximum value of the COE.
4. For every failed scenario, we take the above-mentioned maximum value of the COE, and set the value of the COE for the failed scenario equal to either 1.1 * (maximum value of COE) or 2 * (maximum value of COE)

The reason @omellben proposed this idea is because of the possibility that the feasible optimal solution is near the location with the failed scenario(s).

@eslickj: Can you please let me know what you think? Thank you.

**Comment by eslickj** [comment]
I think we should try it.  I'm not sure what's going to happen with any of these, and the right answer is probably situation dependent.

My fear with this is that there could be a scenario that is far worse than the others.  Since we want to keep the number of scenarios down, we probably won't have scenarios really close together.  My worry is that if the worst scenario fails to solve giving it the highest COE of the remaining scenarios times whatever could still represent a big improvement.  So the trick is to define the multiplier.  You may have a good idea of how much worse the worst generally is than the other scenarios and use that.   It could work out well.

We could also keep historical data fro the scenarios,  then if a scenario fails give it the worst of its own COEs from past runs (maybe times some multiplier).

I think with the custom objective function, implementing most of these things is pretty easy.  No reason not to try anything that seams reasonable.

**Comment by eslickj** [comment]
I should also point out that I suspect the worst scenario may be the most likely to fail to solve. The fact that it's bad probably indicates its farther from what we expect to happen.  It could be on the edge of feasibility.

**Comment by fsoepyan** [comment]
To help me get started writing the objective function code in FOQUS, I am using the calculateObj function from the file foqus_lib\framework\optimizer\problem.py

**Comment by ksbeattie** [comment]
@fsoepyan is working on this.

**Comment by fsoepyan** [comment]
I have been testing the ideas proposed by @omellben and @eslickj regarding the objective function, and this work is ongoing.

**Comment by fsoepyan** [comment]
After testing the ideas proposed by @omellben and @eslickj, the three of us reached a consensus regarding the objective function we should use for the OUU.

Our next step is to test the objective function, where the OUU is run using all 25 scenarios, instead of only 5 scenarios.

**Comment by ksbeattie** [comment]
This issue (or a possible fix) is related to #676

**Comment by fsoepyan** [comment]
I am currently running the OUU with all 25 samples in my computer, with the idea that @omellben and @eslickj proposed.

This run is being performed using FOQUS and Cloud.

**Comment by fsoepyan** [comment]
We ran across an issue: #715

**Comment by fsoepyan** [comment]
Status update:
- @boverhof fixed the issue in #715
- We are running the OUU using 5 workers on Cloud
- We are running the OUU using all 25 scenarios, instead of 5 scenarios
- I am currently testing the idea proposed by @eslickj and @omellben for the objective function
- The first OUU run is successful
- I decided to do a second OUU run, where I slightly modified the objective function
- In other words, this test is to determine the "best" objective function
- No modifications on Cloud or the FOQUS code are needed

**Comment by ksbeattie** [comment]
@fsoepyan still testing this.

**Comment by fsoepyan** [comment]
- I think it's safe to say that @omellben, @eslickj, and I have reached a consensus regarding what the objective function should be to achieve this goal, where if a scenario in an iteration fails, we penalize the iteration instead of throwing out the entire iteration.
- I think we can close this issue.
- Then, I will open another issue so that I can upload the new OUU file to GitHub.
- How does that sound?

**Comment by fsoepyan** [comment]
I believe that once PR #744 is merged, we can close this issue.
